utils/reconnecting_udp_conn_test.go

```diff
@@ -0,0 +1,478 @@
+	args := m.Called(network, laddr, raddr)
+
+	a0 := args.Get(0)
+	if a0 == nil {
+		return (*net.UDPConn)(nil), args.Error(1)
+	}
+
+	return a0.(*net.UDPConn), args.Error(1)
+}
+
+func newUDPListener() (net.PacketConn, error) {
+	return net.ListenPacket("udp", "127.0.0.1:0")
+}
+
+func newUDPListenerOnPort(port int) (net.PacketConn, error) {
+	return net.ListenPacket("udp", fmt.Sprintf("127.0.0.1:%d", port))
+}
+
+func newUDPConn() (net.PacketConn, *net.UDPConn, error) {
+	mockServer, err := newUDPListener()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	addr, err := net.ResolveUDPAddr("udp", mockServer.LocalAddr().String())
+	if err != nil {
+		mockServer.Close()
+		return nil, nil, err
+	}
+
+	conn, err := net.DialUDP("udp", nil, addr)
+	if err != nil {
+		mockServer.Close()
+		return nil, nil, err
+	}
+
+	return mockServer, conn, nil
+}
+
+func assertSockBufferSize(t *testing.T, expectedBytes int, conn *net.UDPConn) bool {
+	fd, _ := conn.File()
+	bufferBytes, _ := syscall.GetsockoptInt(int(fd.Fd()), syscall.SOL_SOCKET, syscall.SO_SNDBUF)
+
+	// The linux kernel doubles SO_SNDBUF value (to allow space for bookkeeping overhead) when it is set using setsockopt(2), and this doubled value is returned by getsockopt(2)
+	// https://linux.die.net/man/7/socket
+	if runtime.GOOS == "linux" {
+		return assert.GreaterOrEqual(t, expectedBytes*2, bufferBytes)
+	}
+
+	return assert.Equal(t, expectedBytes, bufferBytes)
+}
+
+func assertConnWritable(t *testing.T, conn udpConn, serverConn net.PacketConn) {
+	expectedString := "yo this is a test"
+	_, err := conn.Write([]byte(expectedString))
+	require.NoError(t, err)
+
+	var buf = make([]byte, len(expectedString))
+	err = serverConn.SetReadDeadline(time.Now().Add(time.Second))
+	require.NoError(t, err)
+
+	_, _, err = serverConn.ReadFrom(buf)
+	require.NoError(t, err)
+	require.Equal(t, []byte(expectedString), buf)
+}
+
+func waitForCallWithTimeout(call *mock.Call) bool {
+	called := make(chan struct{})
+	call.Run(func(args mock.Arguments) {
+		close(called)
+	})
+
+	var wasCalled bool
+	// wait at most 100 milliseconds for the second call of ResolveUDPAddr that is supposed to fail
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond*100)
+	select {
+	case <-called:
+		wasCalled = true
+	case <-ctx.Done():
+		fmt.Println("timed out")
+	}
+	cancel()
+
+	return wasCalled
+}
+
+func waitForConnCondition(conn *reconnectingUDPConn, condition func(conn *reconnectingUDPConn) bool) bool {
+	var conditionVal bool
+	for i := 0; i < 10; i++ {
+		conn.connMtx.RLock()
+		conditionVal = condition(conn)
+		conn.connMtx.RUnlock()
+		if conditionVal || i >= 9 {
+			break
+		}
+
+		time.Sleep(time.Millisecond * 10)
+	}
+
+	return conditionVal
+}
+
+func newMockUDPAddr(port int) (*net.UDPAddr, error) {
+	var buf = make([]byte, 4)
+	// random is not seeded to ensure tests are deterministic (also doesnt matter if ip is valid)
+	_, err := rand.Read(buf)
+	if err != nil {
+		return nil, err
```

Comment:
nit: could be simplified by taking `t` argument and doing `require.NoError(t, err)`